### PR TITLE
LPD-26175 - language selectors on private layouts with no equivalent public layout returns 404

### PR DESCRIPTION
**Pull request by wanderlast**
Hi @smottal,

https://liferay.atlassian.net/browse/LPD-26175, tied to LPP

This is a fix for a issue arising from the change made in [LPD-20228](https://liferay.atlassian.net/browse/LPD-20228) where we switched from the language selector using the PLID to having it use the layoutID and groupID due to security issues related to using the PLID. When this happens in servicePreAction, we assume that we can also grab whether it's a privateLayout from the request. This isn't necessarily true and therefore it will resolve as 'false' and will attempt to find a public layout with that layoutID and groupID. If the private layout does not have an equivalent public layout this leads to a 404. To fix this, we're simply adding it as an additional parameter to ensure we grab the correct layout later.

### CODE REVIEW

**Comment by smottal** [comment]
Hey @wanderlast , can you provide a test for this bug and resolution? Thx

**Comment by wanderlast** [comment]
ci:test:sf

**Comment by wanderlast** [comment]
ci:test:playwright

**Comment by smottal** [comment]
@wanderlast can you check the test because it's failing. Thx
The code itself looks good

**Comment by wanderlast** [comment]
ci:test:sf

**Comment by wanderlast** [comment]
ci:test:playwright

**Comment by wanderlast** [comment]
ci:test:sf

**Comment by wanderlast** [comment]
ci:test:playwright

**Comment by wanderlast** [comment]
ci:test:playwright

**Comment by wanderlast** [comment]
ci:test:sf

File: modules/apps/portal/portal-user-locale-options-web/src/main/resources/META-INF/resources/dynamic_include/user_locale_options.jsp [hunk 1]
```diff
@@ -22,7 +22,7 @@ Locale userLocale = user.getLocale();
 		<c:if test="<%= LanguageUtil.isAvailableLocale(themeDisplay.getSiteGroupId(), user.getLocale()) || (PortalUtil.isGroupControlPanelPath(themeDisplay.getURLCurrent()) && LanguageUtil.isAvailableLocale(userLocale)) %>">
 			<clay:link
 				cssClass="d-block"
-				href='<%= themeDisplay.getPathMain() + "/portal/update_language?redirect=" + URLCodec.encodeURL(themeDisplay.getURLCurrent()) + "&groupId=" + themeDisplay.getScopeGroupId() + "&languageId=" + user.getLanguageId() + "&layoutId=" + layout.getLayoutId() + "&persistState=false&showUserLocaleOptionsMessage=false" %>'
+				href='<%= themeDisplay.getPathMain() + "/portal/update_language?redirect=" + URLCodec.encodeURL(themeDisplay.getURLCurrent()) + "&groupId=" + themeDisplay.getScopeGroupId() + "&languageId=" + user.getLanguageId() + "&layoutId=" + layout.getLayoutId() + "&persistState=false&showUserLocaleOptionsMessage=false" + "&privateLayout=" + layout.isPrivateLayout() %>'
```
Comment (smottal) [inline modules/apps/portal/portal-user-locale-options-web/src/main/resources/META-INF/resources/dynamic_include/user_locale_options.jsp:25]: can you move &privateLayout parameter between the persistState one and showUserLocalOptionsMessage?

File: modules/test/playwright/tests/users-admin-web/userPersonalSite.spec.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+import {expect, mergeTests} from '@playwright/test';
+
+import {loginTest} from '../../fixtures/loginTest';
+import {usersAndOrganizationsPagesTest} from '../../fixtures/usersAndOrganizationsPagesTest';
+import {liferayConfig} from '../../liferay.config';
+
+export const test = mergeTests(loginTest(), usersAndOrganizationsPagesTest);
+
+test('LPD-26175 check that language selector works on private pages', async ({
```
Comment (smottal) [inline modules/test/playwright/tests/users-admin-web/userPersonalSite.spec.ts:14]: can you please move the check -> Check?

File: modules/test/playwright/tests/users-admin-web/userPersonalSite.spec.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+	await userPersonalSitePage.addLanguageSelectorToPage();
+	await userPersonalSitePage.switchLanguages(
+		'deutsch-Deutschland',
+		'Select a Language'
+	);
+
+	await expect(page).toHaveURL(new RegExp(`.+/user/.+`));
+
+	await page.goto(
```
Comment (smottal) [inline modules/test/playwright/tests/users-admin-web/userPersonalSite.spec.ts:28]: just for info: after this redirect to the update_language page what does it should do?

File: portal-web/docroot/html/common/themes/user_locale_options.jsp [hunk 4]
```diff
@@ -25,14 +25,14 @@ String currentURL = PortalUtil.getCurrentURL(request);
 		</div>
 
 		<c:if test="<%= LanguageUtil.isAvailableLocale(themeDisplay.getSiteGroupId(), user.getLocale()) %>">
-			<aui:a cssClass="d-block" href='<%= themeDisplay.getPathMain() + "/portal/update_language?redirect=" + URLCodec.encodeURL(currentURL) + "&groupId=" + themeDisplay.getScopeGroupId() + "&languageId=" + user.getLanguageId() + "&layoutId=" + layout.getLayoutId() + "&persistState=false&showUserLocaleOptionsMessage=false" %>'>
+			<aui:a cssClass="d-block" href='<%= themeDisplay.getPathMain() + "/portal/update_language?redirect=" + URLCodec.encodeURL(currentURL) + "&groupId=" + themeDisplay.getScopeGroupId() + "&languageId=" + user.getLanguageId() + "&layoutId=" + layout.getLayoutId() + "&persistState=false&showUserLocaleOptionsMessage=false" + "&privateLayout=" + layout.isPrivateLayout() %>'>
```
Comment (smottal) [inline portal-web/docroot/html/common/themes/user_locale_options.jsp:28]: can you move the parameter to be alphabetically sorted?

**Review by smottal — CHANGES_REQUESTED** [review]

File: modules/test/playwright/tests/users-admin-web/userPersonalSite.spec.ts (hunk 3 above)
Comment (wanderlast) [reply modules/test/playwright/tests/users-admin-web/userPersonalSite.spec.ts:28]: The page.goto navigates to the home page and forces the user's language to return to English.

**Comment by wanderlast** [comment]
ci:test:relevant

File: modules/apps/portal/portal-user-locale-options-web/src/main/resources/META-INF/resources/dynamic_include/user_locale_options.jsp [hunk 5]
```diff
@@ -22,7 +22,7 @@ Locale userLocale = user.getLocale();
 		<c:if test="<%= LanguageUtil.isAvailableLocale(themeDisplay.getSiteGroupId(), user.getLocale()) || (PortalUtil.isGroupControlPanelPath(themeDisplay.getURLCurrent()) && LanguageUtil.isAvailableLocale(userLocale)) %>">
 			<clay:link
 				cssClass="d-block"
-				href='<%= themeDisplay.getPathMain() + "/portal/update_language?redirect=" + URLCodec.encodeURL(themeDisplay.getURLCurrent()) + "&groupId=" + themeDisplay.getScopeGroupId() + "&languageId=" + user.getLanguageId() + "&layoutId=" + layout.getLayoutId() + "&persistState=false&showUserLocaleOptionsMessage=false" %>'
+				href='<%= themeDisplay.getPathMain() + "/portal/update_language?redirect=" + URLCodec.encodeURL(themeDisplay.getURLCurrent()) + "&groupId=" + themeDisplay.getScopeGroupId() + "&languageId=" + user.getLanguageId() + "&layoutId=" + layout.getLayoutId() + "&persistState=false" + "&privateLayout=" + layout.isPrivateLayout() + &showUserLocaleOptionsMessage=false" %>'
```
Comment (smottal) [inline modules/apps/portal/portal-user-locale-options-web/src/main/resources/META-INF/resources/dynamic_include/user_locale_options.jsp:25]: I think a " is missing before &showUserLocaleOptionsMessage

**Review by smottal — CHANGES_REQUESTED** [review]

**Review by smottal — APPROVED** [review]

File: modules/apps/portal/portal-user-locale-options-web/src/main/resources/META-INF/resources/dynamic_include/user_locale_options.jsp (hunk 5 above)
Comment (wanderlast) [reply modules/apps/portal/portal-user-locale-options-web/src/main/resources/META-INF/resources/dynamic_include/user_locale_options.jsp]: Thanks for catching that, I've fixed it.

**Comment by smottal** [comment]
ci:test:sf

**Comment by smottal** [comment]
ci:stop

**Comment by smottal** [comment]
ci:test:relevant

**Comment by smottal** [comment]
ci:forward

**Pull request closed without merging**